src/Microsoft.Tye.Core/ApplicationFactory.cs

```diff
@@ -21,227 +21,257 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
+                        {
+                            throw new CommandException("Executable services do not support volumes.");
+                        }
+                        else if (service is ExternalServiceBuilder)
+                        {
+                            throw new CommandException("External services do not support volumes.");
+                        }
+                        else
+                        {
+                            throw new CommandException("Unable to determine service type.");
+                        }
                     }
                 }
-            }
 
-            foreach (var configIngress in config.Ingress)
-            {
-                var ingress = new IngressBuilder(configIngress.Name!);
-                ingress.Replicas = configIngress.Replicas ?? 1;
+                foreach (var configIngress in config.Ingress)
+                {
+                    var ingress = new IngressBuilder(configIngress.Name!);
+                    ingress.Replicas = configIngress.Replicas ?? 1;
 
-                builder.Ingress.Add(ingress);
+                    root.Ingress.Add(ingress);
 
-                foreach (var configBinding in configIngress.Bindings)
-                {
-                    var binding = new IngressBindingBuilder()
+                    foreach (var configBinding in configIngress.Bindings)
                     {
-                        Name = configBinding.Name,
-                        Port = configBinding.Port,
-                        Protocol = configBinding.Protocol ?? "http",
-                    };
-                    ingress.Bindings.Add(binding);
-                }
+                        var binding = new IngressBindingBuilder()
+                        {
+                            Name = configBinding.Name,
+                            Port = configBinding.Port,
+                            Protocol = configBinding.Protocol ?? "http",
+                        };
+                        ingress.Bindings.Add(binding);
+                    }
 
-                foreach (var configRule in configIngress.Rules)
-                {
-                    var rule = new IngressRuleBuilder()
+                    foreach (var configRule in configIngress.Rules)
                     {
-                        Host = configRule.Host,
-                        Path = configRule.Path,
-                        Service = configRule.Service,
-                    };
-                    ingress.Rules.Add(rule);
+                        var rule = new IngressRuleBuilder()
+                        {
+                            Host = configRule.Host,
+                            Path = configRule.Path,
+                            Service = configRule.Service,
+                        };
+                        ingress.Rules.Add(rule);
+                    }
                 }
             }
 
-            return builder;
+            return root;
```

Comment:
One thing I expected to see here:

```yaml
# a.yaml
services:
- name: foo
  project: ....
- name: bar
   include: ../some-path/b.yaml
...
# b.yaml
services:
- name: bar
  project: ....
- name: redis
   image: redis
```

Currently right now you'll get an application with 3 services, `foo`, `bar`, and `redis`. Everything is visible to everything else, meaning that `redis` is a dependency of `foo` and `bar` - they both get access to `redis`'s environment variables.

I expected to see this dependency information represented somewhere, and the binding processing code updated to handle it.

Reply:
This seems pretty subtle, but it also kinda important. The reason why that it becomes possible to have a dependency in code on something you don't have in `tye.yaml`. So if we inject env-vars for a service you don't claim to reference, you've effectively written working code for something undocumented.

Reply:
This might be worth more debate or discussion, but I was assuming the services you *didn't* include to be private dependencies

Reply:
Yeah that is an important distinction that we should fix. I'll look into how to make it so environment variables are only injected for direct dependents.

Reply:
This change is actually a bit more difficult than I thought. I'm going to make a follow up PR with this change after merging this.